### ISSUE TITLE
[BUG]: Card Element doesn't bind to existing Elements provider when using promisified Stripe

### ISSUE THREAD

**Issue by oleg-codaio** [stripe/react-stripe-js#300]
### What happened?

After [upgrading to React Stripe.js](https://github.com/stripe/react-stripe-js/blob/master/docs/migrating.md), we roughly set up our app like this:

```tsx
function StripeWrapper({children}) {
  const stripePromise = useMemo(() => loadStripe('stripe-key'), []);
  return <Elements stripe={stripePromise}>{children}</Elements>;
}

function CreditCardForm() {
  const elements = useElements();

  const onChange = () => {
    const cardElement = elements.getElement(CardElement);
    console.log(cardElement); // undefined!
  };

  return <CardElement onChange={onChange} />
}

function PaymentPage() {
  return (
    <StripeWrapper>
      <CreditCardForm />
    </StripeWrapper>
  );
}
```

The issue here is that `elements.getElement(CardElement)` wouldn't return the handle to the element, and in fact `elements._elements` was empty. I noticed that it looked like when the `CardElement` was being created, it was binding to a elements instance with a different ID.

This is a very simplified version of what we're rendering and it was hard for me to debug this in depth since it looks like Stripe Elements are hot loaded and minified. But I did find a workaround, which was to wait for the `Promise<Stripe | null>` to resolve before rendering anything. Once I passed in `Stripe` directly into `<Elements>`, I no longer had this issue, so I wonder if there's some kind of race condition that only happens with promises.

I was following [this page](https://stripe.com/docs/stripe-js/react) to set everything up. Am I doing something wrong here or is there some kind of bug? Thanks.

### Environment

Chrome on macOS, Version 102.0.5005.61

**Comment by tylersmith-stripe** [comment]
Hey @oleg-codaio,

At first glance, I don't believe this is an issue with the react-stripe-js package. In general, creating side effects in `useMemo` is an anti-pattern, and should be avoided. This suggestion is straight from the React docs: https://reactjs.org/docs/hooks-reference.html#usememo

> Remember that the function passed to useMemo runs during rendering. **Don’t do anything there that you wouldn’t normally do while rendering. For example, side effects belong in useEffect, not useMemo.**

With that said, I'd suggest either:
- Follow [the example from the docs](https://stripe.com/docs/stripe-js/react#elements-provider), placing the `loadStripe` call as a local variable in the file you need it in, or as a singleton export if you need it across multiple files.
- Converting this `useMemo` approach to a combination of `useEffect` and `useState`.

I hope this help you with your issue!

**Comment by oleg-codaio** [comment]
Hey @tylersmith-stripe, that's a fair call on `useMemo` - but unfortunately switching it to a singleton doesn't make it go away. In fact, our actual code is using a custom hook based on `useRef` that initializes it one time and I just switched to `useMemo` for simplicity in describing this issue.

I also just tried to repro this again by moving `stripePromise` out into a global:

```ts
const stripePromise = loadStripe('stripe-key');

function StripeWrapper({children}) {
  return <Elements stripe={stripePromise}>{children}</Elements>;
}
```

And can confirm the issue still persists. Just wanted to flag it in case the issue is in the Stripe library - as I mentioned, it doesn't appear that Elements is open source (AFAICT) so it's a bit hard for me to debug.

**Comment by tylersmith-stripe** [comment]
Okay, interesting! I'll take another look and see if I can add additional clarification here.

**Comment by tylersmith-stripe** [comment]
@oleg-codaio Can you point me to any documentation that is describing the `elements.get(ElementsComponent)` style? Afaik, this would only work by passing the string name of the target Element, `elements.get('card')`.

Here's a link to our JS API Ref for [elements.getElement](https://stripe.com/docs/js/elements_object/get_element).

**Comment by oleg-codaio** [comment]
Sure, here are two places I've seen this syntax:

1. https://github.com/stripe/react-stripe-js/blob/master/docs/migrating.md#after-4
2. https://stripe.com/docs/payments/accept-card-payments?platform=web&ui=elements&html-or-react=react#web-submit-payment (line 23)

FWIW changing to `elements.getElement('card')` didn't help - likely since if I inspect `elements`, the `_elements` property is just equal to `[]` (but there is some other orphaned instance of `elements` that gets created that does contain an item; that second instance only gets created if I'm passing in a promise instead of `Stripe` into the `<Elements>` component).

**Comment by tylersmith-stripe** [comment]
Thank you for sharing those examples. I've confirmed that the `elements.getElement(CardElement)` syntax does indeed work!

Here's a CodeSandbox with your example code (sans `useMemo`), and the console log does appear to be returning the Element instance. https://codesandbox.io/s/little-water-ujevsf?file=/src/App.js

Can you try to recreate your error scenario in a reduced CodeSandbox?

<img width="574" alt="image" src="https://user-images.githubusercontent.com/72463083/172302184-ba0be2c0-e87b-4976-8142-d2b1f878b764.png">

**Comment by oleg-codaio** [comment]
Alright well I spent a good number of hours building a simple repro yesterday: https://codepen.io/vaskevich/pen/wvyELNy

And then realized that it only happens on version 1.8.0 and not 1.8.1, and was fixed by #297. I should have checked the issue history before filing a new one. 🤦🏻 

So FYI @bmathews-stripe / @christopher-stripe - you should now have a repro of #296 with that CodePen (might need to refresh, but notice how it prints out two instances and later binds to the wrong one).

**Comment by bmathews-stripe** [comment]
Ah heck, glad you got the versioning issue sorted. Thanks so much for putting together this repro though! This will be super helpful in figuring why the react effect lifecycle wasn't matching our expectations.
